### ISSUE TITLE
v9.3 planning

### ISSUE THREAD

**Issue by kisvegabor** [kisvegabor/lvgl#6763]
**Update:**
Feature freeze week until June1, Release is planned to June 3. 

### Introduce the problem

**Goals**
I have 2 goals for 2024 Q3 and Q4:
1. Make LVGL more appealing for MPUs: ready-to-use projects, Linux drivers, more docs, heavier features 
2. Improve the developer experience: docs and CI review, more examples, etc

**Schedule**
Targeting 2025 Q1

### Proposal

In light of that, I would like to focus on these features and updates:
**Features**
- [ ] ~Blurring https://github.com/lvgl/lvgl/issues/6018 (deferred as low priority)~
- [x] Tiled rendering https://github.com/lvgl/lvgl/pull/6761
- [x] OpenGL rendering https://github.com/lvgl/lvgl/issues/6017
- [x] Multi-touch (scroll, zoom, rotate) and mouse wheel scroll https://github.com/lvgl/lvgl/issues/6014
- [ ] ~Widget caching  https://github.com/lvgl/lvgl/issues/4910 (deferred in lack time)~
- [x] SVG https://github.com/lvgl/lvgl/issues/6022
- [x] Improve the span widget https://github.com/lvgl/lvgl/issues/6019
- [x] Double and triple click: https://github.com/lvgl/lvgl/pull/6187
- [x] Font manager https://github.com/lvgl/lvgl/issues/6806

**Docs**
- [ ] Review and complete the docs https://github.com/lvgl/lvgl/issues/6671 (partly done)
- [x] Yocto and Build root docs
- [x] Add the docs page for all chip vendor partners
- [ ] Write detailed tutorials about the important features (cancelled)
- [ ] Add more examples where needed (deferred to the next release, collecting ideas now)
 
**Developer experience**
- [x] Have 5 repos for MPU-based dev boards
- [x] Have repos for all the officially supported boards 
- [x] Control the number of open issues and PRs (Adding workflows internally to have better focus on issues/PRs)
- [x] API unification (for the editor and binding generation)
- [ ] ~Clean up the CI https://github.com/lvgl/lvgl/issues/5396 (deferred n lack of time)~
- [ ] ~Separate lv_conf.h to platform and UI-specific parts https://github.com/lvgl/lvgl/issues/6010 (cancelled)~
- [x] Have a demo for MPUs (Full HD)

---

Leave a comment if you would like to see other features too in v9.3. 

Also, let us know if you are interested in contributing to any of these features. 

cc @liamHowatt @XuNeo @FASTSHIFT @etag4048

**Comment by XuNeo** [comment]
Should we also consider optimizing flash usage (maybe RAM usage too)?

**Comment by kisvegabor** [comment]
What do you mean by that in more detail? Disabling modules? Or a general review?

**Comment by XuNeo** [comment]
> What do you mean by that in more detail? Disabling modules? Or a general review?

Last time I tried to disable widgets to reduce flash usage on my `luavgl` project, it fails to compile. Widgets should be disabled automatically when the one it depends on is disabled. 

Disable software renderers not used also saves flash according to https://github.com/lvgl/lvgl/pull/6313. 

We may need to investigate what other modules can be disabled. A white paper to conclude the methods would be great.

**Comment by kisvegabor** [comment]
> Last time I tried to disable widgets to reduce flash usage on my luavgl project, it fails to compile.

Bindings are really a special case as the linker can't eliminate the unused functions. 

> Widgets should be disabled automatically when the one it depends on is disabled.

I wouldn't disable them automatically because it might result in hard to debug config issues. Instead I'd add `#error`s about some other module is required for this one.

**Comment by xlbao123** [comment]
I hope LVGL can become more modular by providing a plugin mechanism, placing all non-core plugins in a unified directory. The advantages of this approach are: 
- it would make module updates and development easier
- allow for easier removal of unnecessary modules
- reduce the need for macros
- simplify the code in lv_init

Some modules that come to mind are:
- File I/O Manager: Similar to the current lv_fs_drv.
- Resource Manager: Supporting such as resource_file, resource_image, resource_font, etc., and managing all resources uniformly with support for cache management algorithms.
- Font Manager.
- Screen Transition Plugin.
- Various image formats/decoders, SVG, Lottie, etc., as independent plugins.
- Gesture Plugin for touch.

**Comment by kisvegabor** [comment]
@xlbao123 Now the `others` (optional extra modules) and `libs` (3rd party lib support) has their own folder. How would a better folder structure look like in your opinion?

**Comment by xlbao123** [comment]
> @xlbao123 Now the `others` (optional extra modules) and `libs` (3rd party lib support) has their own folder. How would a better folder structure look like in your opinion?

My idea is to compile the entire core into` liblvgl.a`, and all other modules as independent libraries in the form of plugins, such as 
- gesture plugin:` liblvgl-plugin-gesture.a`
- fio manager plugin: ` liblvgl-plugin-io-posix.a`, ` liblvgl-plugin-image-memfs.a`, ` liblvgl-plugin-image-vfs.a` (FS like fatfs/littlefs are integrated via vfs)
- image plugin:` liblvgl-plugin-image-lodepng.a`, `liblvgl-plugin-image-svg.a`, ...
- font manager plugin: `liblvgl-plugin-font-freetype.a`, `liblvgl-plugin-font-binfont.a`, ...
- render plugin: `liblvgl-plugin-render-swrender.a`, `liblvgl-plugin-render-vglite.a`, `liblvgl-plugin-render-dma2d.a`. (Since some renderers might partially use software rendering for certain apis, it might be necessary to separate out `liblvgl-swrender.a` as well.)

**Each plugin would have a plugin_name, version, unified init/deinit interfaces, and plugin-specific options.**

This approach would eliminate a large number of macros,  each platform would simply link to different libraries, and making it easy to develop new plugins as well.

**Comment by xlbao123** [comment]
I have a question, and I'm not sure if I understand it correctly. I see that lvgl uses a refr timer to periodically check for dirty rectangles. For optimal performance, I set this timer to 60Hz.

- On certain pages, such as one where the content only refreshes once per second, the timer will frequently wake the MCU from sleep to check for dirty rectangles, and then it will go back to sleep if there's no update. This can waste a lot of power. How can I make sure that the redraw is triggered only when there is a change in the page data?

- Once a redraw is complete, the MCU waits for the next refresh timer period to start rendering the next frame. How can I save this time (so that as soon as one frame is rendered, and the next frame's buffer is ready and there is a redraw event, the next frame starts rendering immediately)?

**Comment by kisvegabor** [comment]
> Each plugin would have a plugin_name, version, unified init/deinit interfaces, and plugin-specific options.
This approach would eliminate a large number of macros, each platform would simply link to different libraries, and making it easy to develop new plugins as well.

How would the code look like? Would you put all the plugins into the same `lvgl` monorepo?

> On certain pages, such as one where the content only refreshes once per second, the timer will frequently wake the MCU from sleep to check for dirty rectangles, and then it will go back to sleep if there's no update. This can waste a lot of power. How can I make sure that the redraw is triggered only when there is a change in the page data?

It's already fixed (https://github.com/lvgl/lvgl/commit/d824618366e474eef57081d647031965e6edc136), please try with the latest master.

> Once a redraw is complete, the MCU waits for the next refresh timer period to start rendering the next frame. How can I save this time (so that as soon as one frame is rendered, and the next frame's buffer is ready and there is a redraw event, the next frame starts rendering immediately)?

Can you just simple select 120Hz instead of 60? :slightly_smiling_face: If there is nothing to do the refresh timer will be paused (with the fix mentioned above)

**Comment by xlbao123** [comment]
> > Each plugin would have a plugin_name, version, unified init/deinit interfaces, and plugin-specific options.
> > This approach would eliminate a large number of macros, each platform would simply link to different libraries, and making it easy to develop new plugins as well.
> 
> How would the code look like? Would you put all the plugins into the same `lvgl` monorepo?

I will try to describe it using some code (in the code below, I’ve added support for the dynamic library, but I think this may not be necessary).

`lv_plugin.h`
```
#ifndef LV_PLUGIN_H
#define LV_PLUGIN_H

struct lv_plugin_state;

typedef struct
{
    char *filename;
    void *dll_handle;
    int valid;
    struct lv_plugin_state info;
} lv_plugin_handle_t;

#define FMT_PLUGIN_VERSION(major, minor, build) (major << 24 | minor << 16 | build)
#define GET_MAJOR(v) (((v) >> 24) & 0xff)
#define GET_MINOR(v) (((v) >> 16) & 0xff)
#define GET_BUILD(v) ((v) & 0xffff)

typedef enum
{
    SERVICE_NUMBER_EARLY = -100,
    SERVICE_NUMBER_NORMAL = 0,
    SERVICE_NUMBER_LATE = 100,
} lv_service_number_t;

typedef uint32_t lv_version_t;

typedef struct lv_plugin_state
{
    const char *name;
    void *plugin_data;
    int (*init)(char *options, void *pdata);
    void (*destory)(void *pdata);
    lv_version_t version;
    int service_number;
} lv_plugin_state_t;

typedef int (*lv_plugin_create_func_t)(lv_plugin_state_t *state);

#if defined(LV_ENABLE_DYNAMIC_PLUGINS)

#define LV_PLUGIN_CREATE_FUNC_STR "lv_plugin_create"

#if defined(_WIN32)
#define LV_PLUGIN_CREATE_FUNC DLLExport lv_plugin_create
#else
#define LV_PLUGIN_CREATE_FUNC lv_plugin_create
#endif /* _WIN32 */

#else

// For static plugin deployments, define a single entry point for the plugin list to be provided.
const lv_plugin_create_func_t *lv_get_plugins();

#endif /* LV_ENABLE_DYNAMIC_PLUGINS */

#endif /* LV_PLUGIN_H */
```

`lv_plugin.c`
```
#include <lv_plugin.h>

typedef struct
{
    // ...
} plugin_private_data_t;

static char *plugin_opts[] = {
#define OPT0 0
    "opt0",
#define OPT1 1
    "opt1",
    NULL,
};

static int
lv_lodepng_init(char *args, void *_pdata)
{
    char *value = NULL;
    while (args && *args != '\0')
    {
        switch (lv_getsubopt((char **)&args, plugin_opts, &value))
        {
        case OPT0:
            break;
        case OPT1:
            break;
        }
    }
    plugin_private_data_t *pdata = (plugin_private_data_t *)_pdata;
    // ...

    return 0;
}

static void
lv_lodepng_destory(void *_pdata)
{
    plugin_private_data_t *pdata = (plugin_private_data_t *)_pdata;
    // ...
    lv_free(pdata);
}

#if !defined(LV_ENABLE_DYNAMIC_PLUGINS)
#define LV_PLUGIN_CREATE_FUNC lv_plugin_image_lodepng
#endif

int LV_PLUGIN_CREATE_FUNC(lv_plugin_state_t *state)
{
    memset(state, 0, sizeof(*state));
    state->version = FMT_PLUGIN_VERSION(1, 0, 0);
    state->name = lv_strdup("lodepng");
    state->init = lv_lodepng_init;
    state->destory = lv_lodepng_destory;
    state->service_number = SERVICE_NUMBER_NORMAL;

    state->plugin_data = lv_calloc(1, sizeof(plugin_private_data_t));

    return 0;
}
```

`lv_init.c`
```
static char *get_plugin_options(const char *plugin_name, char *const *options, int option_count)
{
    // Extract the options for a specific plugin into a getsubopt() style string.
}

#ifdef LV_ENABLE_DYNAMIC_PLUGINS
static int init_plugins(char *const *options, int option_count)
{
    // Load the .so/.dll files from the specified directory.
}

#else
static int init_plugins(char *const *options, int option_count)
{
    lv_plugin_create_func_t plugin_create;
    lv_plugin_state_t *plugin_state;
    char *suboptions;

    const lv_plugin_create_func_t *plugins = lv_get_plugins();
    if (plugins == NULL)
    {
        return 0;
    }

    for (i = 0; plugins[i] != NULL; i++)
    {
        plugin_create = plugins[i];

        lv_plugin_handle_t *plugin = (lv_plugin_handle_t *)lv_calloc(1, sizeof(lv_plugin_handle_t));
        plugin_state = plugin->info;

        ret = plugin_create(plugin_state);
        if (ret != 0 || plugin_state->init == NULL)
        {
            continue;
        }

        suboptions = get_plugin_options(plugin_state->name, options, option_count);

        LV_LOG_INFO("Initialize plugin [%s (%d.%d.%d)] [%s]",
                    plugin_state->name, GET_MAJOR(plugin_state->version),
                    GET_MINOR(plugin_state->version), GET_BUILD(plugin_state->version),
                    (suboptions) ? suboptions : "");

        ret = plugin_state->init(suboptions, plugin_state->plugin_data);
        if (suboptions != NULL)
        {
            lv_free(suboptions);
        }

        if (ret == 0)
        {
            // add to plugin list
        }
    }

    return 0;
}

#endif

int lv_init(char *const *options, int option_count)
{
    // ...
    init_plugins(options, option_count);
    // ...
}
```

`main.c`
```
static const lv_plugin_create_func_t lv_plugins[] = {
    lv_plugin_image_lodepng,
    // ...
    NULL,
};

const lv_plugin_create_func_t *lv_get_plugins()
{
    return lv_plugins;
}

int main(int argc, char **argv)
{
#define OPTIONS_NUM (32)
    char *options[OPTIONS_NUM];
    int option_count = 0;

    options[option_count++] = "lodepng";   ///> plugin name
    options[option_count++] = "opt0=val0"; ///> plugin option

    options[option_count++] = "lodepng";
    options[option_count++] = "opt1=val1";

    assert(option_count <= OPTIONS_NUM);
    lv_init(options, option_count);

    // ...
}
```

1. `lv_plugins` may makes the code look ugly. However, there are ways to optimize this, we can link plugin entry points into a specific section, just like modules in linux kernel.

2. As the features of LVGL continue to grow, one of the advantages of doing this is that we can gradually eliminate more macros: 
    - Feature toggles can be replaced by whether the corresponding plugin is linked or not.
    - For the configuration of specific plugin, if it's a compile-time option, it can be controlled internally by macros within the plugin; if it's a runtime option, it can be replaced by `plugin_option`.

3. Separating the modules is just the first step, and a further step would be interface isolation: avoiding direct calls from the core to plugin interfaces, and also preventing plugins from calling each other's interfaces. Another challenge is to gradually separate some modules from liblvgl, making the core more streamlined.

4. As for the repo management of plugins, I think some core plugins can still be placed in a unified directory within the lvgl repo, while others, especially those not closely tied to the LVGL version, can be maintained in independent repos.

**Comment by kisvegabor** [comment]
Sorry for the late reply. I see the potential value in this approach, but let's wait a little before jumping into it. I'd like to see if there is a real need for such a change.

**Comment by stachurr** [comment]
It would be nice to allow access to fragments on the stack by index. Currently, fragments on the stack can only be found if each fragment has a unique container via `lv_fragment_manager_find_by_container`, which is not guaranteed/enforced.

Here's code that would accomplish this:
```
lv_fragment_t * lv_fragment_manager_get_by_index(lv_fragment_manager_t * manager, size_t index)
{
    LV_ASSERT(manager);
    size_t idx = 0;
    lv_fragment_managed_states_t * states;
    LV_LL_READ(&manager->attached, states) {
        if (idx++ == index) return states->instance;
    }
    return NULL;
}
```

**Comment by kisvegabor** [comment]
To be honest I haven't used the fragments in real projects yet, but your request seems reasonable. Please send a PR. :slightly_smiling_face:

**Comment by stachurr** [comment]
Maybe it is my misunderstanding of the fragments concept, but wouldn't it be desirable to have only one container for all associated fragments? This is what makes them useful, no? I will make a PR in the near future. Thanks!

**Comment by nicusorcitu** [comment]
@kisvegabor do you have a much clear date for release 9.3? I see is Q1 but a more precise date will help. Thanks.

**Comment by kisvegabor** [comment]
Probably the end of Q1 (end of March) is the most realistic. Is there a deadline on your end that we should align with?

**Comment by nicusorcitu** [comment]
I thought it would be released sooner in January, we have a release freeze in February, so definitely 9.3 will not be on time.

**Comment by kisvegabor** [comment]
There are a few feature that we should have in v9.3 and they are not ready yet:
- OpenGL rendering: Basics are  working, but there are some issues on embedded devices. @liamHowatt and @EDGEMTech-GabrielC are tracking it
- Multi-touch (scroll, zoom, rotate) and mouse wheel scroll: The basics are there, but needs to be completed with more feature support @etag4048 and @EDGEMTech-GabrielC 
- SVG: BAsics are working the image decoder part is not ready. @onecoolx is working on it 

The others can slip to the next release. So I think realistically we can have a feature freeze at the end of February. 🙁 Is it a big issue for you? If yes, we can figure out something.

**Comment by cshilwant** [comment]
> Make LVGL more appealing for MPUs: ready-to-use projects, Linux drivers, more docs, heavier features

@kisvegabor Are there any plans to add Gstreamer rendering support (similar to FFmpeg) as MPUs heavily rely on gstreamer for multimedia streaming?

**Comment by kdschlosser** [comment]
Think we are going to hit the target date for this release? the end of Q1 2025 is in < 2 weeks.

**Comment by kisvegabor** [comment]
> @kisvegabor Are there any plans to add Gstreamer rendering support (similar to FFmpeg) as MPUs heavily rely on gstreamer for multimedia streaming?

Yes, it just came up a few weeks ago, but it's not scheduled on our end yet. 

> Think we are going to hit the target date for this release? the end of Q1 2025 is in < 2 weeks.

I'll check the ongoing things and start wrapping it up. It might slip to beginning of April.

**Comment by liamHowatt** [comment]
Reminder to revert or keep this breaking change before v9.3: https://github.com/lvgl/lvgl/pull/7310#issuecomment-2623245065

**Comment by liamHowatt** [comment]
Note to self and others to ensure this is fixed before v9.3: #7797

**Comment by grenhand** [comment]
Hello, may I ask if the upcoming V9.3 release will support a font manager?

**Comment by kisvegabor** [comment]
By font manager, do you mean this? https://github.com/lvgl/lvgl/issues/6806

**Comment by grenhand** [comment]
> 字体管理器，您指的是这个吗？[#6806](https://github.com/lvgl/lvgl/issues/6806)

Yes.

**Comment by kisvegabor** [comment]
@FASTSHIFT we have discussed that the font manager will be extended beyond FreeType to support the built-in fonts too.  (https://github.com/lvgl/lvgl/issues/6806). Will have time work on it in the near future?

**Comment by FASTSHIFT** [comment]
> [@FASTSHIFT](https://github.com/FASTSHIFT) we have discussed that the font manager will be extended beyond FreeType to support the built-in fonts too. ([#6806](https://github.com/lvgl/lvgl/issues/6806)). Will have time work on it in the near future?

Let's start here: https://github.com/lvgl/lvgl/pull/8038

**Comment by grenhand** [comment]
By the way, for FreeType, which version do you recommend? I'm on the latest version, but I'm getting compilation errors.

![Image](https://github.com/user-attachments/assets/eb2a9c73-b6ae-4566-9510-2a3d2d95567e)

**Comment by FASTSHIFT** [comment]
> By the way, for FreeType, which version do you recommend? I'm on the latest version, but I'm getting compilation errors.
> 
> ![Image](https://github.com/user-attachments/assets/eb2a9c73-b6ae-4566-9510-2a3d2d95567e)

The compilation error is because the freetype library is not correctly linked. Please install it according to the [documentation](https://docs.lvgl.io/master/details/libs/freetype.html#for-unix) and execute:
```bash
cd build
cmake .. -DLV_USE_FREETYPE=ON
make -j
```

**Comment by grenhand** [comment]
> > By the way, for FreeType, which version do you recommend? I'm on the latest version, but I'm getting compilation errors.
> > ![Image](https://github.com/user-attachments/assets/eb2a9c73-b6ae-4566-9510-2a3d2d95567e)
> 
> The compilation error is because the freetype library is not correctly linked. Please install it according to the [documentation](https://docs.lvgl.io/master/details/libs/freetype.html#for-unix) and execute:
> 
> cd build
> cmake .. -DLV_USE_FREETYPE=ON
> make -j

Thanks,i fixed it by adding ftsystem.c to the project.

**Comment by grenhand** [comment]
I used lv_freetype_font_create to create a new font. When I applied LV_FREETYPE_FONT_STYLE_ITALIC, it worked, but LV_FREETYPE_FONT_STYLE_BOLD didn't. What am I missing?

**Comment by nicusorcitu** [comment]
hi @kisvegabor 
When is the release planned?

**Comment by sukesh-ak** [comment]
@kisvegabor 
What are the plans for `lv_observer`. 
Is it going to be part of lvgl core or are there plans to have a replacement of it?

I am about to start a project and was looking to build UI API for it.

**Comment by FASTSHIFT** [comment]
> I used lv_freetype_font_create to create a new font. When I applied LV_FREETYPE_FONT_STYLE_ITALIC, it worked, but LV_FREETYPE_FONT_STYLE_BOLD didn't. What am I missing?

`LV_FREETYPE_FONT_STYLE_BOLD` support is also incomplete and temporarily unavailable.

**Comment by cosmindanielradu19** [comment]
> Are there any open items on NXP's side?

Yes, we have changes for both VGLite an G2D.
PRs will be opened soon.

**Comment by kisvegabor** [comment]
Just commented in https://github.com/lvgl/lv_font_conv/pull/134

**Comment by GorgonMeducer** [comment]
@kisvegabor #8113
1. The cmsis-pack is done. Please do NOT merge as we should wait until the last minute of the cooling period. 
2. I will update the arm-2d adaptor to introduce more acceleration opportunities. (by end of this week)

**Comment by kisvegabor** [comment]
Thank you Gabriel!

**Comment by GorgonMeducer** [comment]
@kisvegabor 
In #7244 the LVGL intrinsics support was damaged...
We get issues like this:

```c
C:/Users/gabriel/AppData/Local/Arm/Packs/LVGL/lvgl/9.3.0/src/draw/sw/lv_draw_sw_img.c(293): error: use of undeclared identifier 'draw_unit'
  293 |                                                   draw_unit,        /* target buffer, buffer width, buffer height, buffer stride */
      |                                                   ^
C:/Users/gabriel/AppData/Local/Arm/Packs/LVGL/lvgl/9.3.0/src/draw/sw/lv_draw_sw_img.c(561): error: use of undeclared identifier 'blend_area'
  561 |         if(LV_RESULT_INVALID == LV_DRAW_SW_RGB565_RECOLOR(dest_buf, blend_area, color, mix)) {
      |                                                                     ^
C:/Users/gabriel/AppData/Local/Arm/Packs/LVGL/lvgl/9.3.0/src/draw/sw/lv_draw_sw_img.c(600): error: use of undeclared identifier 'blend_area'
  600 |         if(LV_RESULT_INVALID == LV_DRAW_SW_RGB888_RECOLOR(dest_buf, blend_area, color, mix, cf_final)) {
      |                                                                     ^
C:/Users/gabriel/AppData/Local/Arm/Packs/LVGL/lvgl/9.3.0/src/draw/sw/lv_draw_sw_img.c(600): error: use of undeclared identifier 'cf_final'
  600 |         if(LV_RESULT_INVALID == LV_DRAW_SW_RGB888_RECOLOR(dest_buf, blend_area, color, mix, cf_final)) {
```

**Comment by GorgonMeducer** [comment]
@kisvegabor I have finished updating the arm2d support and the cmsis-pack is ready to review.

By the way, I have saw at least 15 last minutes PRs. I have updated the cmsis-pack to catchup with them.

If you have more, please let me know. I need to update the cmsis-pack until the last minute.

**Comment by GorgonMeducer** [comment]
@kisvegabor Three weeks passed. Any plan for the release? Thank you.

**Comment by kisvegabor** [comment]
Sorry, there is only one key thing left: https://github.com/lvgl/lvgl/pull/8204
Once it's merged we are good for the feature freeze week.

**Comment by GorgonMeducer** [comment]
I am about to go on vacation on Friday and will not have stable access to GitHub until June.  
@kisvegabor

**Comment by kisvegabor** [comment]
Thank you! ❤️

**Comment by GorgonMeducer** [comment]
CMSIS-Pack is up-to-date. @kisvegabor

**Comment by faxe1008** [comment]
@kisvegabor  Found and fixed an issue regarding `LV_USE_THEME_MONO` with `I1` color format displays in the past several people in Zephyr stumbled upon. Would be lovely to have this in 9.3 if its not a hassle:
https://github.com/lvgl/lvgl/pull/8343

**Comment by kisvegabor** [comment]
@faxe1008 merged, thank you! 

@GorgonMeducer Just fixed a conflict, merging CMSIS soon. One more review is needed.

**Comment by kisvegabor** [comment]
V9.3 was released yesterday. It was also pushed to all library/component registries. 

As no crisis was reported so far, I close this issue. 


Thank you very much for the hard of you all! ❤️

**Comment by plan44** [comment]
In lack of a better place I add this comment here: what v9.3 now starts to do with XML is something I needed back in 2019 - dynamically configure lvgl UIs in json. By now latest [p44utils](https://github.com/plan44/p44utils/tree/luz) supports that for lvgl v9.2, including styles, screens, events - the latter can just directly execute [p44script](https://plan44.ch/p44-techdocs/en/script_ref/) (a js-like script tailored at (home) automation tasks) with web-style `onxxx="p44script…"` handlers. Just thought I should mention this, it's certainly niche but maybe still of interest to some.

**Comment by kisvegabor** [comment]
Thank you for sharing!
